fun/middleware/oauth2_step.py

```diff
@@ -0,0 +1,51 @@
+"""Middleware to inspect authorize oauth2 request"""
+import re
+from urllib import quote
+
+from django.conf import settings
+from django.shortcuts import resolve_url
+from django.http import HttpResponseRedirect, QueryDict
+from django.utils.six.moves.urllib.parse import urlparse, urlunparse
+
+authorize_path = re.compile(r"/oauth2/authorize/")
+
+class Oauth2Step(object):
+    """
+    When Moodle initiate an Oauth2 request, the state variable in the querystring
+    contains itself a querystring. On the redirect on the login page when a user is 
+    not connected, Django set in a next variable the initial request path and then redirect on this
+    request path once connected. But the state is unquoted and splited in multiple variable in the querystring...
+    To fix this, first we store the initial query string in the request object and then we rebuild
+    the next query string by quoting once again the state to ensure it will not be splited
+    once redirected.
+    """
+   
+    def process_request(self, request):
+        """
+        The initial querystring in store in the request to be use later
+        in the process_response.
+        """
+        if ("state" in request.GET 
+            and not request.user.is_authenticated() 
+            and authorize_path.match(request.path_info)):
+            request._initial_qs = request.GET.copy()
+
+            return None
```

Comment:
```suggestion
        return None
```
There is an indentation issue ?

Reply:
No but Yes :-D